File: day-12-golang-python/fundamental_data.go
// data_engineer_fundamental.go

package main

import (
	"encoding/csv"
	"fmt"
	"log"
	"os"
	"strconv"
)

// 1. Membaca CSV
func loadCSV(path string) ([][]string, error) {
	file, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	reader := csv.NewReader(file)
	return reader.ReadAll()
}

// 2. Membersihkan data (hapus baris yang ada field kosong)
func cleanData(data [][]string) [][]string {
	var cleaned [][]string
	for i, row := range data {
		empty := false
		for _, field := range row {
			if field == "" {
				empty = true
				break
			}
		}
		if !empty || i == 0 { // selalu simpan header
			cleaned = append(cleaned, row)
		}
	}
	return cleaned
}

// 3. Transformasi (contoh: kolom "amount" dikali 1.1, hasil ke kolom "amount_usd")
func transformData(data [][]string) [][]string {
	header := append(data[0], "amount_usd")
	result := [][]string{header}

	amountIdx := -1
	for i, col := range data[0] {
		if col == "amount" {
			amountIdx = i
			break
		}
	}

	if amountIdx == -1 {
		log.Println("Kolom 'amount' tidak ditemukan, skip transformasi")
		return data
	}

	for i, row := range data {
		if i == 0 {
			continue // header sudah ditambahkan
		}
		amount, err := strconv.ParseFloat(row[amountIdx], 64)
		if err != nil {
			row = append(row, "0")
		} else {
			usd := amount * 1.1
			row = append(row, fmt.Sprintf("%.2f", usd))
		}
		result = append(result, row)
	}

	return result
}

// 4. Simpan ke CSV
func saveCSV(path string, data [][]string) error {
	file, err := os.Create(path)
	if err != nil {
		return err
	}
	defer file.Close()

	writer := csv.NewWriter(file)
	defer writer.Flush()

	return writer.WriteAll(data)
}

// 5. Pipeline utama
func main() {
	inputPath := "data/raw_data.csv"
	outputPath := "data/cleaned_data.csv"

	data, err := loadCSV(inputPath)
	if err != nil {
		log.Fatalf("Gagal membaca file CSV: %v", err)
	}

	data = cleanData(data)
	data = transformData(data)

	err = saveCSV(outputPath, data)
	if err != nil {
		log.Fatalf("Gagal menyimpan file: %v", err)
	}

	fmt.Println("Pipeline selesai. Data disimpan di", outputPath)
}

File: day-12-golang-python/fundamental_data.py
# data_engineer_fundamental.py

import pandas as pd
import numpy as np

# 1. Membaca file CSV
def load_data(filepath):
    print(f"Loading data from {filepath}...")
    return pd.read_csv(filepath)

# 2. Menampilkan info dasar
def explore_data(df):
    print("\n--- Head of the Data ---")
    print(df.head())
    print("\n--- Info ---")
    print(df.info())
    print("\n--- Descriptive Stats ---")
    print(df.describe())

# 3. Membersihkan data
def clean_data(df):
    print("\nCleaning data...")
    # Contoh: Menghapus duplikat
    df = df.drop_duplicates()
    # Mengisi missing value pada kolom numerik
    for col in df.select_dtypes(include=np.number).columns:
        df[col] = df[col].fillna(df[col].median())
    # Mengisi missing value pada kolom object (string)
    for col in df.select_dtypes(include='object').columns:
        df[col] = df[col].fillna('unknown')
    return df

# 4. Transformasi data sederhana
def transform_data(df):
    print("\nTransforming data...")
    if 'amount' in df.columns:
        df['amount_usd'] = df['amount'] * 1.1  # Contoh konversi nilai
    return df

# 5. Menyimpan hasil
def save_data(df, output_path):
    print(f"\nSaving cleaned data to {output_path}...")
    df.to_csv(output_path, index=False)

# 6. Pipeline utama
def main():
    input_path = 'data/raw_data.csv'     # Ubah path sesuai kebutuhan
    output_path = 'data/cleaned_data.csv'

    df = load_data(input_path)
    explore_data(df)
    df = clean_data(df)
    df = transform_data(df)
    save_data(df, output_path)
    print("\nPipeline selesai.")

if __name__ == '__main__':
    main()
